src/ncp-spinel/SpinelNCPTaskJoinerCommission.cpp

```diff
@@ -0,0 +1,182 @@
+	{
+		strncpy(mProvisioningUrl, provision_url, sizeof(mProvisioningUrl));
+		mProvisioningUrl[strlen(mProvisioningUrl)] = 0;
+	}
+}
+
+void
+nl::wpantund::SpinelNCPTaskJoinerCommission::finish(int status, const boost::any& value)
+{
+	SpinelNCPTask::finish(status, value);
+}
+
+int
+nl::wpantund::SpinelNCPTaskJoinerCommission::vprocess_event(int event, va_list args)
+{
+	int ret = kWPANTUNDStatus_Failure;
+	Data frame;
+
+	EH_BEGIN();
+
+	if (!mInstance->mEnabled) {
+		ret = kWPANTUNDStatus_InvalidWhenDisabled;
+		finish(ret);
+		EH_EXIT();
+	}
+
+	if (mInstance->get_ncp_state() == UPGRADING) {
+		ret = kWPANTUNDStatus_InvalidForCurrentState;
+		finish(ret);
+		EH_EXIT();
+	}
+
+	// Wait for a bit to see if the NCP will enter the right state.
+	EH_REQUIRE_WITHIN(
+		NCP_DEFAULT_COMMAND_RESPONSE_TIMEOUT,
+		!ncp_state_is_initializing(mInstance->get_ncp_state()),
+		on_error
+	);
+
+	if (ncp_state_is_associated(mInstance->get_ncp_state())) {
+		ret = kWPANTUNDStatus_Already;
+		finish(ret);
+		EH_EXIT();
+	}
+
+	// The first event to a task is EVENT_STARTING_TASK. The following
+	// line makes sure that we don't start processing this task
+	// until it is properly scheduled. All tasks immediately receive
+	// the initial `EVENT_STARTING_TASK` event, but further events
+	// will only be received by that task once it is that task's turn
+	// to execute.
+	EH_WAIT_UNTIL(EVENT_STARTING_TASK != event);
+
+	if (mAction)
+	{
+		mInstance->change_ncp_state(ASSOCIATING);
+
+		// Turn off promiscuous mode, if it happens to be on
+		mNextCommand = SpinelPackData(
+				SPINEL_FRAME_PACK_CMD_PROP_VALUE_SET(SPINEL_DATATYPE_UINT8_S),
+				SPINEL_PROP_MAC_PROMISCUOUS_MODE,
+				SPINEL_MAC_PROMISCUOUS_MODE_OFF
+				);
+		EH_SPAWN(&mSubPT, vprocess_send_command(event, args));
+		ret = mNextCommandRet;
+		check_noerr(ret);
+
+		// Now bring up the network by bringing up the interface
+		mNextCommand = SpinelPackData(
+				SPINEL_FRAME_PACK_CMD_PROP_VALUE_SET(SPINEL_DATATYPE_BOOL_S),
+				SPINEL_PROP_NET_IF_UP,
+				true
+				);
+
+		EH_SPAWN(&mSubPT, vprocess_send_command(event, args));
+		ret = mNextCommandRet;
+		require((ret == kWPANTUNDStatus_Ok) || (ret == kWPANTUNDStatus_Already), on_error);
+	}
+
+	frame.clear();
+	frame.append(reinterpret_cast<const uint8_t*>(&mAction), sizeof(mAction));
+
+	if (strlen(mPsk) > 0)
+	{
+		frame.append(reinterpret_cast<const uint8_t*>(mPsk), strlen(mPsk) + 1);
+	}
+
+	if (strlen(mProvisioningUrl) > 0)
+	{
+		frame.append(reinterpret_cast<const uint8_t*>(mProvisioningUrl), strlen(mProvisioningUrl) + 1);
+	}
+	syslog(LOG_INFO, "Joiner commission %s (psk: %s) (provisioning_url %s)", mAction ? "start" : "stop", mPsk, mProvisioningUrl);
+
+	mNextCommand = SpinelPackData(
+			SPINEL_FRAME_PACK_CMD_PROP_VALUE_SET(SPINEL_DATATYPE_DATA_S),
+			SPINEL_PROP_THREAD_JOINER_COMMISSION,
+			frame.data(),
+			frame.size()
+			);
+
+	EH_SPAWN(&mSubPT, vprocess_send_command(event, args));
+	ret = mNextCommandRet;
+	check_noerr(ret);
```

Comment:
how about `require_noerr(ret, on_error)` instead here?
`check_noerr()`  will add a print and continue (commonly used to indicate it's ok if this check fail).